h2o-core/src/test/java/water/TestUtil.java

```diff
@@ -123,32 +124,75 @@ public static void writeFrameToCSV(String fileNameWithPath, Frame h2oframe, bool
     outStream.close();
   }
 
-  @AfterClass
-  public static void checkLeakedKeys() {
-    int leaked_keys = H2O.store_size() - _initial_keycnt;
-    int cnt=0;
-    if( leaked_keys > 0 ) {
-      for( Key k : H2O.localKeySet() ) {
-        Value value = Value.STORE_get(k);
-        // Ok to leak VectorGroups and the Jobs list
-        if( value==null || value.isVecGroup() || value.isESPCGroup() || k == Job.LIST ||
-            // Also leave around all attempted Jobs for the Jobs list
-            (value.isJob() && value.<Job>get().isStopped()) ) {
-          leaked_keys--;
-        } else {
-          System.out.println(k + " -> " + (value.type() != TypeMap.PRIM_B ? value.get() : "byte[]"));
-          if( cnt++ < 10 )
-            System.err.println("Leaked key: " + k + " = " + TypeMap.className(value.type()));
+  @After
+  public void checkLeakedKeys() {
+    performLeakedKeysCheck();
+  }
+
+
+  protected static void performLeakedKeysCheck() {
+    final int leaked_keys = H2O.store_size() - _initial_keycnt;
+    int unreportedKeyCount = leaked_keys;
+    int nonIgnorableKeyCount = leaked_keys;
+    int cnt = 0;
+    Set<Key> localKeySet = new HashSet<>(H2O.localKeySet());
+
+    if (leaked_keys > 0) {
+      for (Key key : H2O.localKeySet()) {
+        final Value keyValue = Value.STORE_get(key);
+        if (isIgnorableKeyLeak(key, keyValue)) {
+          unreportedKeyCount--;
+          nonIgnorableKeyCount--;
+          localKeySet.remove(key);
+          continue;
+        }
+        cnt++;
+        if (keyValue != null && keyValue.isFrame()) {
+          Frame frame = (Frame) key.get();
+          localKeySet.remove(key);
+          unreportedKeyCount--;
+          Log.err(String.format("Leaked frame with key '%s'. This frame contains the following vectors:", frame._key.toString()));
+
+          for (Key vecKey : frame.keys()) {
+            Log.err(String.format("   Vector '%s'. This vector contains the following chunks:", vecKey.toString()));
+            localKeySet.remove(vecKey);
+            unreportedKeyCount--;
+
+            final Vec vec = (Vec) vecKey.get();
+            for (int i = 0; i < vec.nChunks(); i++) {
+              final Key chunkKey = vec.chunkKey(i);
+              Log.err(String.format("       Chunk id %d, key '%s'", i, chunkKey));
+              localKeySet.remove(vec.chunkKey(i));
+              unreportedKeyCount--;
+            }
+          }
+          key.remove();
         }
       }
-      if( 10 < leaked_keys ) System.err.println("... and "+(leaked_keys-10)+" more leaked keys");
+
+      if (!localKeySet.isEmpty()) {
+        Log.err(String.format("%nThere are also %d more leaked keys:", unreportedKeyCount));
+      }
+
+      for (Key key : localKeySet) {
+        unreportedKeyCount--;
```

Comment:
Why do we need to decrement this here? It looks like we don't use it afterwards.

Reply:
Oh, you're right. fixed